topoembedx/classes/hope.py

```diff
@@ -0,0 +1,174 @@
+
+class HOPE:
+    """Higher Order Laplacian Positional Encoder (HOPE) class.
+
+    Parameters
+    ----------
+    dimensions : int, optional
+        Dimensionality of embedding. Defaults to 3.
+    """
+
+    def __init__(self, dimensions: int = 2):
+
+        self.A = []
+        self.ind = []
+        self.dimensions = dimensions
+
+        self._embedding = []
+
+    @staticmethod
+    def _laplacian_pe(A, k: int, return_eigenval: bool = False):
+        """
+        Compute Laplacian Positional Encodings (PE) for a given adjacency matrix.
+
+        Parameters
+        ----------
+        A : numpy.ndarray, shape (n, n)
+            Adjacency matrix representing the graph structure.
+        k : int
+            Number of eigenvectors to consider for Laplacian PE.
+        return_eigenval : bool, optional (default=False)
+            Whether to return the eigenvalues along with PE.
+
+        Returns
+        -------
+        numpy.ndarray or tuple
+            Laplacian Positional Encodings computed from the Laplacian eigenvectors.
+            If return_eigenval is True, returns a tuple (PE, eigenvalues).
+
+        Explanation
+        -----------
+        This function computes Laplacian Positional Encodings (PE) based on the input
+        adjacency matrix and the desired number of eigenvectors. The Laplacian PE is a
+        representation of the graph structure obtained from the Laplacian eigenvectors.
+
+        The Laplacian matrix L is computed using the input adjacency matrix A and then
+        its eigenvectors are calculated. The k smallest non-trivial eigenvectors are
+        selected to form the Laplacian PE.
+
+        If the number of eigenvectors (k) is less than or equal to the number of nodes (n),
+        the remaining dimensions are filled with zeros. The signs of the eigenvectors are
+        randomly flipped to enhance representational capacity.
+
+        Finally, the Laplacian positional encodings are returned for further usage. If
+        return_eigenval is True, eigenvalues are also returned alongside the positional encodings.
+        """
+        n = A.shape[0]
+        # Compute the degree matrix D^-0.5
+        D = sparse.diags(np.squeeze(np.asarray(np.power(np.sum(A, axis=1), -0.5))))
+
+        # Compute the Laplacian matrix L = I - D^-0.5 * A * D^-0.5
+        L = np.eye(A.shape[0]) - D @ A @ D
+
+        # Compute the eigenvectors of L
+        EigVal, EigVec = np.linalg.eig(L)
+
+        # Select the k smallest non-trivial eigenvectors
+        max_freqs = min(n - 1, k)
+        kpartition_indices = np.argpartition(EigVal, max_freqs)[: max_freqs + 1]
+        topk_eigvals = EigVal[kpartition_indices]
+        topk_indices = kpartition_indices[topk_eigvals.argsort()][1:]
+        topk_EigVec = EigVec[:, topk_indices]
+
+        # Randomly flip signs of the eigenvectors
+        rand_sign = 2 * (np.random.rand(max_freqs) > 0.5) - 1.0
+        PE = np.multiply(rand_sign, topk_EigVec.astype(np.float32))
```

Comment:
Avoid uppercase letters in Python, we only made an exception for A, D.
replace with pe or pos_enc